### PR TITLE
Print QueryLog time range

### PR DESCRIPTION
**Pull request by ishmum123**
Also, adding a summary holder for general usage

### CODE REVIEW

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+import java.time.ZonedDateTime;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author shevek
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java:27]: This comment was likely copied from another file by mistake. If so, it can be removed.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnector.java [hunk 2]
```diff
@@ -385,18 +391,20 @@ protected static class TeradataAssessmentLogsJdbcTask extends TeradataLogsJdbcTa
...
-                "L.WarningOnly",
-                "L.WDName"
+            "L.UtilityInfoAvailable",
+            "L.UtilityRowCount",
+            "L.VHLogicalIO",
+            "L.VHLogicalIOKB",
+            "L.VHPhysIO",
+            "L.VHPhysIOKB",
+            "L.WarningOnly",
+            "L.WDName"
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnector.java:401]: Please, revert the formatting change as it breaks the indent compared to the rows above. If it is needed, then I would recommend moving the formatting changes to a separate PR and applying it to all the strings in the `EXPRESSIONS_FOR_ASSESSMENT` array.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/AbstractJdbcTask.java [hunk 3]
```diff
@@ -89,24 +90,33 @@ protected CSVFormat newCsvFormat(@Nonnull ResultSet rs) throws SQLException {
                 // Can we avoid nesting exceptions here?
                 throw new SQLException(
                         new MetadataDumperUsageException("Fatal Error. ResultSet does not have the expected column count: " + headerClass.getEnumConstants().length,
-                                Arrays.asList("If a custom query has been specified please confirm the selected columns match the following: ",
-                                        StringUtils.join(headerClass.getEnumConstants(), ", "))));
+                            Arrays.asList(
+                                "If a custom query has been specified please confirm the selected columns match the following: ",
+                                StringUtils.join(headerClass.getEnumConstants(), ", "))));
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/AbstractJdbcTask.java:95]: Similarly to the previous comment: please, revert the formatting-only change or move it to a separate PR.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/AbstractJdbcTask.java [hunk 4]
```diff
@@ -89,24 +90,33 @@ protected CSVFormat newCsvFormat(@Nonnull ResultSet rs) throws SQLException {
...
         } else {
             format = format.withHeader(rs);
         }
         return format;
     }
 
     @Nonnull
-    protected ResultSetExtractor<Void> newCsvResultSetExtractor(@Nonnull ByteSink sink, @CheckForSigned long count) {
+    protected ResultSetExtractor<Void> newCsvResultSetExtractor(@Nonnull ByteSink sink,
+        @CheckForSigned long count) {
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/AbstractJdbcTask.java:104]: Optional: I would suggest to either keep the parameters in one line, or move each parameter to a separate line (both `sink` and `count`).

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        getEarliestQueryTime(),
+        getLatestQueryTime()
+    );
+    return new Summary(getTotalRows(), interval);
+  }
+
+  protected void updateMetadata(String queryType, long rowCount, ZonedInterval interval) {
+    metadataMap.putIfAbsent(queryType, new Summary(rowCount, interval));
+    metadataMap.computeIfPresent(queryType,
+        (q, metadata) -> metadata.updateMetadata(rowCount, interval));
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java:53]: If `rowCount` passed as a parameter to this `updateMetadata` method is 1, then what would be the value of `Summary#rowCount` field after the `putIfAbsent` and `computeIfPresent` invocations?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnector.java [hunk 6]
```diff
@@ -154,27 +152,35 @@ protected static class TeradataLogsJdbcTask extends AbstractJdbcTask<Void> {
...
+        private final Consumer<Long> rowCountConsumer;
+
 
         public TeradataLogsJdbcTask(@Nonnull String targetPath, SharedState state, String logTable,
-                                    String queryTable, List<String> conditions, ZonedInterval interval) {
-            this(targetPath, state, logTable, queryTable, conditions, interval, Collections.emptyList());
+            String queryTable, List<String> conditions, ZonedInterval interval,
+            Consumer<Long> consumer) {
+            this(targetPath, state, logTable, queryTable, conditions, interval,
+                Collections.emptyList(), consumer);
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnector.java:162]: nit: I would suggest static importing the `emptyList` method.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnector.java [hunk 7]
```diff
@@ -434,15 +442,20 @@ public void addTasksTo(List<? super Task<?>> out, @Nonnull ConnectorArguments ar
         LOG.info("Exporting query log for " + intervals);
         SharedState state = new SharedState();
         for (ZonedInterval interval : intervals) {
-            String file = ZIP_ENTRY_PREFIX + DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(interval.getStartUTC()) + ".csv";
+            String file = ZIP_ENTRY_PREFIX + DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(
+                interval.getStartUTC()) + ".csv";
+            Consumer<Long> consumer = count -> updateMetadata(ZIP_ENTRY_PREFIX, count, interval);
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnector.java:447]: I would suggest to create a method `createMetadataUpdater` and use it in all the places where updateMetadata is used in the lambda currently.

```
Consumer<Long> createMetadataUpdater(String filePrefix, ZonedInterval interval) { ... }
```

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/AbstractJdbcTask.java [hunk 8]
```diff
@@ -89,24 +90,33 @@ protected CSVFormat newCsvFormat(@Nonnull ResultSet rs) throws SQLException {
...
+        return newCsvResultSetExtractor(sink, count, ignored -> {
+        });
+    }
+
+    protected ResultSetExtractor<Void> newCsvResultSetExtractor(@Nonnull ByteSink sink,
+        @CheckForSigned long count, @Nonnull
+    Consumer<Long> countConsumer) {
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/AbstractJdbcTask.java:111]: Please, fix the parameters formatting.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/MetadataDumper.java [hunk 9]
```diff
@@ -292,9 +293,32 @@ public <T> T runChildTask(Task<T> task) throws MetadataDumperUsageException {
...
                 );
+            } else if (connector instanceof AbstractSummaryHolderJdbcConnector) {
+                AbstractSummaryHolderJdbcConnector.Summary summary =
+                    ((AbstractSummaryHolderJdbcConnector) connector).getSummary();
+                long totalRows = summary.rowCount();
+                if (totalRows > 0) {
+                    System.out.println(
+                        "**********************************************************\n"
+                            + "* Logs has been saved to " + outputFile + "\n"
+                            + "* Earliest Row Found at " + summary.earliestInterval().getStart()
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/MetadataDumper.java:304]: I would suggest to use an explicit formatter (e.g. `DateTimeFormatter.ISO_OFFSET_DATE_TIME.withZone(ZoneOffset.UTC)`) for outputting timestamps. What formatter is used by default?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/MetadataDumper.java [hunk 10]
```diff
@@ -292,9 +293,32 @@ public <T> T runChildTask(Task<T> task) throws MetadataDumperUsageException {
...
+                            + "\n"
+                            + "* Latest Row Found at " + summary.latestInterval().getEndInclusive()
+                            + "\n"
+                            + "* Total Number of Rows Dumped " + totalRows + "\n"
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/MetadataDumper.java:308]: nit: In my opinion, the capitalization of first letters is unnecessary. I would suggest to change to `Total number of rows dumped`.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/snowflake/SnowflakeLogsConnector.java [hunk 11]
```diff
@@ -163,11 +163,6 @@ public void addTasksTo(List<? super Task<?>> out, ConnectorArguments arguments)
 
         String queryFormat = newQueryFormat(arguments);
 
-        final int daysToExport = arguments.getQueryLogDays(7);
-        if (daysToExport <= 0)
-            throw new MetadataDumperUsageException("At least one day of query logs should be exported; you specified: " + daysToExport);
-        LOG.info("Log entries within the last {} days will be exported in increments of 1 hour.", daysToExport);
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/snowflake/SnowflakeLogsConnector.java:169]: Why is this code being removed?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/MetadataDumper.java [hunk 12]
```diff
@@ -292,9 +293,32 @@ public <T> T runChildTask(Task<T> task) throws MetadataDumperUsageException {
...
+                            + "**********************************************************"
+                    );
+                } else {
+                    System.out.println(
+                        "**********************************************************\n"
+                            + "* No Rows were Found\n"
```
Comment (frsv) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/MetadataDumper.java:314]: Few suggestions for the case where there are no logs:

- does it make sense to log here a range, i.e `No rows were found in the range xxx - yyy`?
- we maybe want to emphasize this is a warning and prepend `WARNING` to the message
- if there are no logs dumped, the output of the connector is essentially useless. How about leaving a suggestion to increase dumping time range in the message?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * @author shevek
+ */
+public abstract class AbstractSummaryHolderJdbcConnector extends AbstractJdbcConnector {
+
+  @SuppressWarnings("UnusedVariable")
+  private static final Logger LOG = LoggerFactory.getLogger(
+      AbstractSummaryHolderJdbcConnector.class);
+
+  private final Map<String, Summary> metadataMap;
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java:35]: This gives mutable state to a class which did not previously have state. Effectively, the task class becomes non-reusable. Could we please discuss potential solutions to this?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author shevek
+ */
+public abstract class AbstractSummaryHolderJdbcConnector extends AbstractJdbcConnector {
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java:29]: Does this class need to go into the inheritance hierarchy? Given that it has state, might it be better as a Support-style object which is created during the task/connector runtime?

**Review by shevek-google — CHANGES_REQUESTED** [review]

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java (hunk 5 above)
Comment (ishmum123) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/AbstractSummaryHolderJdbcConnector.java]: Nice catch! Thanks

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/snowflake/SnowflakeLogsConnector.java (hunk 11 above)
Comment (ishmum123) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/snowflake/SnowflakeLogsConnector.java:169]: This check is already being done inside the function at github.com/google/dwh-migration-tools/pull/133/files/d6d1173a85e4295443285c74c74e6fd6da865dc0#diff-7dcc102b5538a3c78cee78295c7c87e5ed1698fe369894cc6402e9785f85aaf4L175

**Review by dawidxc — COMMENTED** [review]
Ishmum, I've left some comments, but I've stopped reviewing, because the PR is very big. It contains:
1. introduction of the computation of the span of 2 ZonedIntervals
2. extraction of the nested classes `TeradataAssessmentLogsJdbcTask` and `TeradataLogsJdbcTask` to higher level
3. convenience methods for logging with the stars-decoration
4. formatting changes
5. fix in the SnowflakeLogsConnector
6. introduction of the `Summary` class and usage of this class in `JdbcSelectTask` class

Possibly some more stuff that I've missed. I would like to kindly request splitting these changes to separate PRs. The smaller PR the better. By the size in this context I don't mean the number of lines of code, but a separate feature (i.e. each feature in a separate PR).

**Comment by ishmum123** [comment]
Depends on #183 atm

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/Summary.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+  public static final BinaryOperator<Summary> COMBINER = (s1, s2) -> {
+    if (s1.rowCount() == 0) return s2;
+    if (s2.rowCount() == 0) return s1;
+
+    long rowCount = s1.rowCount() + s2.rowCount;
+
+    Optional<ZonedInterval> optionalInterval1 = s1.interval();
+    Optional<ZonedInterval> optionalInterval2 = s2.interval();
+
+    if (!optionalInterval1.isPresent()) return new Summary(rowCount, optionalInterval2.orElse(null));
```
[the remark this answers is not in the document]
Comment (ishmum123) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/Summary.java]: `isEmpty()` seems to be only available from jdk11 😞
